### PR TITLE
Cadl repo structure guidelines

### PR DESCRIPTION
**Pull request by tjprescott**
Closes https://github.com/Azure/cadl-azure/issues/1691.

Relocated from original PR: https://github.com/Azure/cadl-azure/pull/2021

### CODE REVIEW

**Comment by heaths** [comment]
1. FWIW, this is an exact duplicate at the time of this writing with the guidelines from the old PR.
2. Were you going to transfer over the versioning spec?
3. Do you have plans to copy over comments on the PR, or will you address them there first? There were quite a few open comments.

**Comment by tjprescott** [comment]
@heaths no this has been updated based on our working group meeting from earlier this week. Versioning was split out because we didn't discuss it, so that is still an open question.

File: documentation/cadl-structure-guidelines.md
Comment (tjprescott) [reply documentation/cadl-structure-guidelines.md:74]: Yes. It lists the emitters and configuration settings for the emitters.

**Comment by weshaggard** [comment]
Lets please get https://github.com/Azure/azure-rest-api-specs/blob/main/README.md#directory-structure updated and linked to this new document as well.

File: documentation/cadl-structure-guidelines.md
Comment (mikekistler) [inline documentation/cadl-structure-guidelines.md:45]: Why create directories under "compute" that replicate the structure within "resource-manager"? Expanding this example a bit, I think it would look like:
```
   -> compute
      -> Compute.Management               (mgmt SDK + service)
      -> Compute.Management.Shared        (shared)
      -> Compute.CloudService.Management  (service)
      -> Compute.Diagnostic.Management    (service)
      -> Compute.Disk.Management          (service)
      -> Compute.Gallery.Management       (service)
      -> Compute.Skus.Management          (service)
      -> resource-manager                 (swagger)
        -> Microsoft.Compute
           -> CloudserviceRP
           -> ComputeRP
           -> DiagnosticRP
           -> DiskRP
           -> GalleryRP
           -> Skus
           -> common
```

While the mapping above is relatively obvious, I fear that this opens the door for crazy structures like:

```
   -> widgets
      -> Widgets.Management               (mgmt SDK + service)
      -> Widgets.Cats.Management  (service)
      -> Widgets.Dogs.Management    (service)
      -> Widgets.Horse.Management          (service)
      -> Widgets.Mule.Management       (service)
      -> Widgets.Sheep.Management          (service)
      -> resource-manager                 (swagger)
        -> Microsoft.Widgets
           -> Chevy
           -> Ford
           -> Nissan
           -> Rivian
           -> Tesla
           -> Toyota
```

File: documentation/cadl-structure-guidelines.md
Comment (mikekistler) [inline documentation/cadl-structure-guidelines.md:25]: Can you give some examples of when these would be useful?

I wonder if we want Cadl docs in the azure-rest-api-specs repo that don't describe an Azure REST API.

**Review by mikekistler — COMMENTED** [review]
I think this could work but I'm concerned about having the Cadl files so far away from the corresponding REST API docs in the directory hierarchy.

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:45]: As we chatted offline the goal is to move to a module where data-plane and resource-manager folders become publish only sections and the work happens in the folders directly in the service directory. We know things aren't perfect and there is some risk there as well but we need to make forward progress and we think this is the best balance with all the requirements.

**Comment by weshaggard** [comment]
@tjprescott can you pull in the versioning information back into this document now and also update the contributing md file to point to this new document? Let me know if I can help move this forward so we can get this PR merged in.

File: documentation/cadl-structure-guidelines.md
Comment (dw511214992) [inline documentation/cadl-structure-guidelines.md:87]: Where is the sidecar file? is it in the folder together with `cadl-project.yaml`? Or it's in sdk repository?

File: documentation/cadl-structure-guidelines.md
Comment (raych1) [inline documentation/cadl-structure-guidelines.md:37]: How about adding a root folder say 'CADL' to keep all packages, Compute.Management, Compute.Management.Shared, etc, so that validation tools running in PR pipeline can be easier to filter out this root folder 'CADL' and no need to worry about the dynamically changed folder names?

File: documentation/cadl-structure-guidelines.md
Comment (raych1) [inline documentation/cadl-structure-guidelines.md:104]: Do you have idea of which tool is supposed to calculate the diff between the PR changes and main branch?

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:87]: It should live in the same folder as the cadl-project.yaml file. However as @tjprescott calls out here there doesn't necessarily always need to be a sidecar file and there also might only be a single one as opposed to one per-language. Those details are still be worked out.

File: documentation/cadl-structure-guidelines.md
Comment (raych1) [inline documentation/cadl-structure-guidelines.md:108]: What is this `cadl` folder? Is it the similar thing I comment above for the root folder to keep all CADL files.

File: documentation/cadl-structure-guidelines.md
Comment (leni-msft) [reply documentation/cadl-structure-guidelines.md:45]: Borrowing idea from @heaths (a.k.a. the Git-friendly folder structure, which link can be found on the related issue), and given Cadl is designed to support multiple versions with decorators in single files, we could use a `dev/` root folder, at the same level with `specification/`, to place all Cadl related files. For example:
```
.
├── dev (or vNext?)
│   └── keyvault
│       ├── KeyVault.Certificates
│       ├── KeyVault.Keys
│       ├── KeyVault.Management
│       │   ├── cadl-project.yaml
│       │   ├── examples
│       │   │   ├── CreateKey-example.json
│       │   │   ├── DeleteKey-example.json
│       │   │   ├── GetKey-example.json
│       │   │   └── GetKeys-example.json
│       │   ├── main.cadl
│       │   ├── tests
│       │   └── package.json
│       └── KeyVault.Secrets
└── specification
    └── keyvault
        ├── data-plane
        │   ├── Microsoft.KeyVault
        │   │   ├── preview
        │   │   └── stable
        │   └── readme.md
        └── resource-manager
            ├── Microsoft.KeyVault
            │   ├── preview
            │   └── stable
            └── readme.md
```
In this case, all Cadl related files will be in the `dev/` folder, and it will be always used to define the latest or 'vNext' API specs (so we may name the folder `vNext/`). 
When publishing, Swagger will be generated to the `specification/` folder - all the existing tools and pipeline based on current folder structure won't be affected at all.

Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:45]: We are definitely using some of the dev related things such as versioning in place instead of side-by-side but we don't want to have multiple root level folders as we want the new cadl projects to live inside of the shared service folder so things are closer together.

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:37]: As we chatted we wanted to keep "cadl" out of the folder structure and try to keep this as simple as possible for the 80% of cases where there will only one mgmt and/or one dpg project folder under the service folder.

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:104]: Eventually I'm hoping this will be the cadl tooling itself, probably the swagger emitter set of tooling.

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:108]: I think those are left over from an earlier proposal. It should be specification.

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:103]: It depends on what local build tools we end up using. If it is the standard cadl tools I don't expect them to understand this constraint. However if we can write custom linter rules that the tool runs perhaps we could include it as part of that.

File: documentation/cadl-structure-guidelines.md
Comment (josefree) [reply documentation/cadl-structure-guidelines.md:25]: @tjprescott would you respond Mike's questions? I have the same question.  What do you mean 'SDK spec'?

File: documentation/cadl-structure-guidelines.md
Comment (josefree) [reply documentation/cadl-structure-guidelines.md:37]: in a word, although under the same <rp-name> folder, cadl specs and swagger specs will be in separate sibling folders. while swagger specs are under folders with fixed names as 'data-plane' and 'resource-manager', cadl specs are under folders named after SDK package.  However, for a first-time onboarding resource provider, the SDK package name need to be reviewed and decided in the SDK ArchBoard Review, which happens after API spec is frozen for SDK release.  Did you see the contradiction here? In the case, the RP can only randomly name the folder first, and change the folder name later.    On the other hand, because the folder name and number of folders are not fixed, tooling needs to scan all folders under <rp-name> folder to decide what are cadl folders for which SDK type.  It is doable, but not straightforward enough.

Comment (josefree) [reply documentation/cadl-structure-guidelines.md:37]: Beside api specs, examples are also key artifacts.  Not sure how service should define examples when using cadl for API definition?  And what are folders for examples?

File: documentation/cadl-structure-guidelines.md
Comment (josefree) [inline documentation/cadl-structure-guidelines.md:61]: why management library needs a suffix, while data plane library does not?

File: documentation/cadl-structure-guidelines.md
Comment (josefree) [inline documentation/cadl-structure-guidelines.md:66]: merging into main branch of the azure-rest-api-specs, but still 'unpublished'?  It's against convention we have today - everything that has been merged into main branch of the public repo azure-rest-api-specs are published

File: documentation/cadl-structure-guidelines.md
Comment (josefree) [inline documentation/cadl-structure-guidelines.md:78]: So when APIs are defined in cadl, examples are in JSON without change.

File: documentation/cadl-structure-guidelines.md
Comment (raych1) [reply documentation/cadl-structure-guidelines.md:103]: Yes, that's what lint rule should cover.

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:37]: The folder names don't necessarily map to the exact name of the sdk a lot of times it more closely maps to the resource provider namespace so I don't think that should really be any more of an issue then we had in the past. Also, we should be careful to avoid hardcoding folder locations so that we can rename them if needed later. By the time we do our first PR merge we should have enough confidence in the name that it shouldn't really be an issue.

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:61]: That is a fair question. Regardless we need some way to identify which is which and most of the time the mgmt sdk's often include the name management or resource manager in the names so that is what we chose here as well.

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:66]: This is specifically talking about the cadl packages. They are similar to a JS/TS package and could be published to npm but in the usage scenarios here we don't want these to be published to npm. 

This is different than publishing a swagger spec to main in the repo as that will still be happening.

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:78]: Yes that is the current expectation.

File: documentation/cadl-structure-guidelines.md
Comment (tjprescott) [reply documentation/cadl-structure-guidelines.md:25]: @mikekistler an example might be where a large service spec contains service A and B, but they want an SDK for A and an SDK for B. They could simply extract the SDK shape for A and B from the service spec using the `@client` decorators (and other DPG decorators). Alternatively, they could simply model A and B as separate services and use the first option (service spec and SDK definition are one and the same). 

The mention here isn't to say that this is a common (or recommended) practice, merely that it's a possibility (and is compatible with this proposal).

Comment (tjprescott) [reply documentation/cadl-structure-guidelines.md:25]: An SDK spec means that the service definition lives elsewhere and that the given cadl describes only the shape of an SDK.

**Review by heaths — CHANGES_REQUESTED** [review]

File: documentation/cadl-structure-guidelines.md
Comment (jianyexi) [reply documentation/cadl-structure-guidelines.md:78]: Will the examples have subfolder for each api-version ?

File: documentation/cadl-structure-guidelines.md
Comment (tjprescott) [reply documentation/cadl-structure-guidelines.md:140]: This is merely the "goals" of the versioning section. None of these bullet points are prescriptive. I could actually remove them since I'm not sure they are relevant to a general audience.

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:78]: That is an interesting question. @tjprescott @johanste do you expect that will be the case? My thought is the examples here under the cadl projects will be version be the aggregate of all versions but when we publish them under the swagger paths we will only publish examples for the given version.

File: documentation/cadl-structure-guidelines.md
Comment (raych1) [reply documentation/cadl-structure-guidelines.md:37]: Is there any requirement for data plane(cadl) folder name? such as case convention (pascal case or camal case), namespace convention (does it must include dot separated segments), how many segments of namespace allowed (a.b or a.b.c), etc.  we need to have them documented. Otherwise, user could have different understanding on the rule which will make them inconsistent.
confidentialledger->ConfidentialLedger: two same name with different case is a little confused to me, why should I have two almost duplicated name in paths.

File: documentation/cadl-structure-guidelines.md
Comment (raych1) [inline documentation/cadl-structure-guidelines.md:39]: Can we consider any folder name that includes 'Management' is mgmt cadl? Or 'Management' has to be in the end or be second word in folder name? Can we have this clearly documented?

File: documentation/cadl-structure-guidelines.md
Comment (raych1) [inline documentation/cadl-structure-guidelines.md:66]: Can we mention that how service team is supposed to consume these **unpublished** packages? What is the version control and support level for these packages when they're referenced by other services?

File: documentation/cadl-structure-guidelines.md
Comment (tjprescott) [reply documentation/cadl-structure-guidelines.md:39]: The proposal says that management libraries end with "Management" unless it is a shared management library, in which case it ends in "Management.Shared".

File: documentation/cadl-structure-guidelines.md
Comment (tjprescott) [reply documentation/cadl-structure-guidelines.md:66]: This is covered in the proposal. Specs should not need to be consumed, but shared libraries are consumed using relative file reference syntax.

File: documentation/cadl-structure-guidelines.md
Comment (tjprescott) [reply documentation/cadl-structure-guidelines.md:139]: The service would only need to save v2-preview.1 in Cadl until it is no longer serviced, at which time the service team would remove it, but if that preview is still in service at the same time as  preview.2, then they would both need to be represented. However, preview.1 is _always_ preserved in Git history, so you could always track the commit hash for it and get it that way, which is the process documented for a complete break like that.

**Review by weshaggard — CHANGES_REQUESTED** [review]
Most of the docs look good to me but I'd like to see the cadl stuff at https://github.com/Azure/azure-rest-api-specs/blob/main/README.md#directory-structure updated before merging this.

File: CONTRIBUTING.md
Comment (tjprescott) [inline CONTRIBUTING.md:51]: @weshaggard is this what you want?

File: documentation/cadl-structure-guidelines.md
Comment (tjprescott) [reply documentation/cadl-structure-guidelines.md]: Fixed

**Comment by tjprescott** [comment]
> Most of the docs look good to me but I'd like to see the cadl stuff at https://github.com/Azure/azure-rest-api-specs/blob/main/README.md#directory-structure updated before merging this.

@weshaggard I've updated both README.md and contributing.md

File: README.md
Comment (heaths) [inline README.md:47]: 1. Nit:
```suggestion
6. **'preview' and 'stable' Folders**: This maps to the service/component lifecycle stage: Preview and GA. For example, if a service is in Preview stage, no matter Private Preview or Public Preview, the API specs of the service should be placed in the `preview` folder. If the service is GAed, but a component is in preview, then the API version contains the preview component entity should be placed in the `preview` folder as well.  The `stable` folder should contain API versions of a GAed service and all GAed components. 
```
2. This is only true for swaggers, though, right? Nothing here seems to indicate that Cadls should contain all versions and _not_ be in separate preview and stable folders.

File: CONTRIBUTING.md
Comment (weshaggard) [reply CONTRIBUTING.md:51]: I suspect this isn't the best link for coding style for now lets keep the existing link.

File: README.md
Comment (weshaggard) [inline README.md:40]: We should add a like to your document here.

File: CONTRIBUTING.md
Comment (weshaggard) [inline CONTRIBUTING.md:60]: @josefree should we actually have the directory structure detail inside of here instead of the readme.md directly?

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:37]: I suspect we should use pascal casing with `.` separators very muchs similar to the RP and/or namespace guideline we already have. No real change there.

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:78]: This is something we need to work out but not something to block this PR from merging.

File: documentation/cadl-structure-guidelines.md
Comment (weshaggard) [reply documentation/cadl-structure-guidelines.md:139]: I'm fine with more detailed branch docs but I don't think this is necessarily the document for that detail. @heaths if you would like to start another PR for review that would be great.

File: CONTRIBUTING.md
Comment (weshaggard) [inline CONTRIBUTING.md:51]:
```suggestion
REST APIs for Azure should be defined using the [OpenAPI v2](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md) or [Cadl](https://aka.ms/cadl/rpaas-start)format.
```

File: README.md
Comment (weshaggard) [inline README.md:40]:
```suggestion
    RP folders may contain resoure manager or data plane Cadl specs. Cadl is a language for describing cloud service APIs and generating other API description languages, client and service code, documentation, and other assets. Explore more by visiting the tutorial in the CADL repo: [Cadl tutorial](http://aka.ms/cadlTutorial). You can also ask questions for providing feedback in the internal Teams channel [Cadl Discussion](https://teams.microsoft.com/l/channel/19%3a906c1efbbec54dc8949ac736633e6bdf%40thread.skype/Cadl%2520Discussion%2520%25F0%259F%2590%25AE?groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47).
    
    For more information about the structure of Cadl files in the repo see [Cadl repo structure](https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/cadl-structure-guidelines.md).
```

File: README.md
Comment (weshaggard) [inline README.md:102]:
```suggestion
|        +---Playfab
|        |   \---cadl-project.yaml  
|        |   \---main.cadl  
```

**Review by weshaggard — APPROVED** [review]

**Review by heaths — APPROVED** [review]

File: documentation/cadl-structure-guidelines.md
Comment (raych1) [reply documentation/cadl-structure-guidelines.md:66]: In cross service reference scenario, will shared libraries version upgrade break existed reference scenario? Will this be restricted by breaking change rule?

**Pull request merged**
